Modules/xSharePoint/DSCResources/MSFT_xSPUserProfileProperty/MSFT_xSPUserProfileProperty.psm1

```diff
@@ -0,0 +1,388 @@
+        [parameter(Mandatory = $false)]  [System.string ] $MappingDirection ,
+        [parameter(Mandatory = $false)]  [System.Int32] $Length ,
+        [parameter(Mandatory = $false)]  [System.Int32] $DisplayOrder ,
+        [parameter(Mandatory = $false)]  [System.Boolean] $IsEventLog ,
+        [parameter(Mandatory = $false)]  [System.Boolean] $IsVisibleOnEditor ,
+        [parameter(Mandatory = $false)]  [System.Boolean] $IsVisibleOnViewer ,
+        [parameter(Mandatory = $false)]  [System.Boolean] $IsUserEditable ,
+        [parameter(Mandatory = $false)]  [System.Boolean] $IsAlias ,
+        [parameter(Mandatory = $false)]  [System.Boolean] $IsSearchable,
+        [parameter(Mandatory = $false)]  [System.Boolean] $UserOverridePrivacy ,
+        [parameter(Mandatory = $false)]  [System.string ] $TermStore ,
+        [parameter(Mandatory = $false)]  [System.string ] $TermGroup ,
+        [parameter(Mandatory = $false)]  [System.string ] $TermSet ,
+        [parameter(Mandatory = $false)] [System.Management.Automation.PSCredential] $InstallAccount
+    )
+
+    #note for integration test: CA can take a couple of minutes to notice the change. don't try refreshing properties page. go through from a fresh "flow" from Service apps page :)
+
+    Write-Verbose -Message "Creating user profile property $Name"
+    $test = $PSBoundParameters
+    $result = Invoke-xSharePointCommand -Credential $InstallAccount -Arguments $test -ScriptBlock {
+        $params = $args[0]
+        #region Validating parameter combinations
+        if( ($params.ContainsKey("TermSet")  -or $params.ContainsKey("TermGroup") -or $params.ContainsKey("TermSet") ) -and
+            ($params.ContainsKey("TermSet")  -and $params.ContainsKey("TermGroup") -and $params.ContainsKey("TermSet") -eq $false ) 
+            )
+        {
+            throw "You have to provide all 3 parameters Termset, TermGroup and TermStore when providing any of the 3."
+        }
+
+        #what if combination property type + termstore isn't possible?
+        if($params.ContainsKey("TermSet")  -and (@("string","stringmultivalue").Contains($params.Type.ToLower()) -eq $false)  ){
+            throw "Only String and String Maultivalue can use Termsets"
+        }
+        #endregion 
+        #region setting up objects 
+        $ups = Get-SPServiceApplication -Name $params.UserProfileService -ErrorAction SilentlyContinue 
+ 
+        If ($null -eq $ups)
+        {
+            return $null
+        }
+        
+        $caURL = (Get-SpWebApplication  -IncludeCentralAdministration | ?{$_.IsAdministrationWebApplication -eq $true }).Url
+        $context = Get-SPServiceContext  $caURL 
+
+        $userProfileConfigManager = new-object Microsoft.Office.Server.UserProfiles.UserProfileConfigManager($context)
+        if($null -eq $userProfileConfigManager)
+        {   #if config manager returns when ups is available then isuee is permissions
+            throw "account running process needs permissions"
```

Comment:
Can we be more specific with the "permissions" that are required here? Just so the error is more meaningful. What are the permissions anyway? Are they permissions we can grant and then revoke when we are done if needed?